components/sync15/src/telemetry.rs

```diff
@@ -0,0 +1,645 @@
+    applied: u32,
+
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    failed: u32,
+
+    #[serde(rename = "newFailed")]
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    new_failed: u32,
+
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    reconciled: u32,
+}
+
+impl EngineIncoming {
+    pub fn new() -> Self {
+        Self {
+            applied: 0,
+            failed: 0,
+            new_failed: 0,
+            reconciled: 0,
+        }
+    }
+
+    fn is_empty(inc: &Option<Self>) -> bool {
+        match inc {
+            Some(a) => a.applied == 0 && a.failed == 0 && a.new_failed == 0 && a.reconciled == 0,
+            None => true,
+        }
+    }
+
+    pub fn applied(&mut self, n: u32) {
+        self.applied += n;
+    }
+
+    pub fn failed(&mut self, n: u32) {
+        self.failed += n;
+    }
+
+    pub fn new_failed(&mut self, n: u32) {
+        self.new_failed += n;
+    }
+
+    pub fn reconciled(&mut self, n: u32) {
+        self.reconciled += n;
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct EngineOutgoing {
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    sent: usize,
+
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    failed: usize,
+}
+
+impl EngineOutgoing {
+    pub fn new() -> Self {
+        EngineOutgoing { sent: 0, failed: 0 }
+    }
+
+    pub fn sent(&mut self, n: usize) {
+        self.sent += n;
+    }
+
+    pub fn failed(&mut self, n: usize) {
+        self.failed += n;
+    }
+}
+
+// One engine's sync.
+#[derive(Debug, Serialize)]
+pub struct Engine {
+    name: String,
+
+    #[serde(flatten)]
+    when_took: Stopwatch,
+
+    #[serde(skip_serializing_if = "EngineIncoming::is_empty")]
+    incoming: Option<EngineIncoming>,
+
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    outgoing: Vec<EngineOutgoing>, // one for each batch posted.
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(rename = "failureReason")]
+    failure: Option<SyncFailure>,
+}
+
+impl Engine {
+    pub fn new(name: &str) -> Self {
+        Self {
+            name: name.to_string(),
+            when_took: Stopwatch::new(),
+            incoming: None,
+            outgoing: Vec::new(),
+            failure: None,
+        }
+    }
+
+    pub fn incoming(&mut self, inc: EngineIncoming) {
+        assert!(self.incoming.is_none());
+        self.incoming = Some(inc);
+    }
+
+    pub fn outgoing(&mut self, out: EngineOutgoing) {
+        self.outgoing.push(out);
+    }
+
+    pub fn failure(&mut self, failure: SyncFailure) {
+        assert!(self.failure.is_none());
+        self.failure = Some(failure);
+    }
+
+    fn finished(&mut self) {
+        self.when_took = self.when_took.finished();
+    }
+}
+
+#[cfg(test)]
+mod engine_tests {
+    use super::*;
+
+    #[test]
+    fn test_engine() {
+        let mut e = Engine::new("test_engine");
+        e.finished();
+        assert_json(&e, json!({"name": "test_engine", "when": 0.0}));
+    }
+
+    #[should_panic]
```

Comment:
Is the panic you expect here that you think assert_json will fail? Or some part of serializing. If the former, then I'd honestly maybe prefer an assert_not_json too or something. If the latter, then maybe just serialize it without the assert_json.

(In general I don't love #[should_panic] for this reason, it's not always clear which bit should panic, and you can have it pass for the wrong reasons easily).